### PR TITLE
Ensure cache key only contains valid characters

### PR DESCRIPTION
**Pull request by lekoala**
targeted at 5.0 from https://github.com/silverstripe/silverstripe-framework/pull/10890

### CODE REVIEW

**Comment by GuySartorelli** [comment]
Please don't have two separate PRs for the same thing - please close either this pr or the original pr.

**Review by michalkleiner — APPROVED** [review]

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
How did you reach this list of valid characters?
The line directly above this for example includes a hyphen in the key - is that invalid? If so we shouldn't be putting it there in the first place. If it _is_ valid - we should include it, along with any other valid characters, in the regex.

**Comment by lekoala** [comment]
@GuySartorelli $this->getComposerLockPath() returns an absolute path with / or \\. I just did a quick fix, obviously it's also possible to pick another cache key, but i wanted to minimize any impact

**Comment by GuySartorelli** [comment]
Is your aim here then simply to remove the slashes? If so, please _only_ remove the slashes. As I mentioned in my comment above we're explicitly adding a `-` into the key and then immediately removing it, with this PR, which seems a bit silly. We should either have a wider range of allowed characters that includes characters we are _intentionally_ adding to the key, or we should explicitly only remove characters we know to be problematic.

**Comment by lekoala** [comment]
@GuySartorelli agreed, it doesn't make a lot of sense. i've put everything in one line (filtering invalid chars from the lock path) and replaced the "-" by a "_" which is what is PSR 6 requires. "-" is not officially supported, so it's probably best to avoid it.

**Comment by lekoala** [comment]
Also fixes https://github.com/silverstripe/silverstripe-framework/issues/10490

File: src/Core/Manifest/VersionProvider.php [hunk 1]
```diff
@@ -80,7 +80,7 @@ public function getVersion()
      */
     public function getModuleVersion(string $module): string
     {
-        $key = sprintf('%s-%s', $this->getComposerLockPath(), $module);
+        $key = sprintf('%s_%s', preg_replace("/[^A-Za-z0-9]/", '', $this->getComposerLockPath()), $module);
```
Comment (lozcalver) [inline src/Core/Manifest/VersionProvider.php]: `$module` needs to be included in the `preg_replace()` here, as the module name contains slashes (e.g. `silverstripe/framework`)

**Review by lozcalver — CHANGES_REQUESTED** [review]

File: src/Core/Manifest/VersionProvider.php (hunk 1 above)
Comment (lekoala) [reply src/Core/Manifest/VersionProvider.php]: ah yes.. then we might as well remove the sprintf altogether, that's what i just did in the last commit

File: src/Core/Manifest/VersionProvider.php [hunk 2]
```diff
@@ -44,7 +44,7 @@ class VersionProvider
      */
     public function getVersion()
     {
-        $key = sprintf('%s-%s', $this->getComposerLockPath(), 'all');
+        $key = preg_replace("/[^A-Za-z0-9]/", '', $this->getComposerLockPath() . 'all');
```
Comment (michalkleiner) [inline src/Core/Manifest/VersionProvider.php]: Not sure how many people will ever read the output, but we can use a delimiter not to have it all munged together. Not too fussed if this gets added in the end or not.
```suggestion
        $key = preg_replace("/[^A-Za-z0-9]/", '_', $this->getComposerLockPath() . '_all');
```

File: src/Core/Manifest/VersionProvider.php [hunk 3]
```diff
@@ -80,7 +80,7 @@ public function getVersion()
      */
     public function getModuleVersion(string $module): string
     {
-        $key = sprintf('%s-%s', $this->getComposerLockPath(), $module);
+        $key = preg_replace("/[^A-Za-z0-9]/", '', $this->getComposerLockPath() . $module);
```
Comment (michalkleiner) [inline src/Core/Manifest/VersionProvider.php]:
```suggestion
        $key = preg_replace("/[^A-Za-z0-9]/", '_', $this->getComposerLockPath() . '_' . $module);
```

**Comment by GuySartorelli** [comment]
This looks okay to me. @michalkleiner @robbieaverill do either of you have outstanding concerns?

**Review by robbieaverill — APPROVED** [review]
LGTM

**Review by michalkleiner — APPROVED** [review]
Fine with me

**Review by GuySartorelli — APPROVED** [review]

**Pull request merged**